keps/sig-storage/1432-volume-health-monitor/README.md

```diff
@@ -674,8 +679,180 @@ _This section must be completed when targeting alpha to a release._
+  Ideally, this should be a metric. Operations against the Kubernetes API (e.g.,
+  checking if there are objects with field X set) may be a last resort. Avoid
+  logs or events for this purpose.
+  An operator can check the metrics `csi_sidecar_operations_seconds`,
+  Container Storage Interface operation duration with gRPC error code status
+  total. It is reported from CSI external-health-monitor-controller and
+  csi-external-health-monitor-agent sidecars.
+
+* **What are the SLIs (Service Level Indicators) an operator can use to determine
+the health of the service?**
+  - [ ] Metrics
+    - Metric name: csi_sidecar_operations_seconds
+    - [Optional] Aggregation method:
+    - Components exposing the metric: csi-external-health-monitor-controller and csi-external-health-monitor-agent sidecars
+  - [ ] Other (treat as last resort)
+    - Details:
+
+* **What are the reasonable SLOs (Service Level Objectives) for the above SLIs?**
+  At a high level, this usually will be in the form of "high percentile of SLI
+  per day <= X". It's impossible to provide comprehensive guidance, but at the very
+  high level (needs more precise definitions) those may be things like:
+  - per-day percentage of API calls finishing with 5XX errors <= 1%
+  - 99% percentile over day of absolute value from (job creation time minus expected
+    job creation time) for cron job <= 10%
+  - 99,9% of /health requests per day finish with 200 code
+
+* **Are there any missing metrics that would be useful to have to improve observability
+of this feature?**
+  Describe the metrics themselves and the reasons why they weren't added (e.g., cost,
+  implementation difficulties, etc.).
+
+### Dependencies
+
+_This section must be completed when targeting beta graduation to a release._
+
+* **Does this feature depend on any specific services running in the cluster?**
+  Think about both cluster-level services (e.g. metrics-server) as well
+  as node-level agents (e.g. specific version of CRI). Focus on external or
+  optional services that are needed. For example, if this feature depends on
+  a cloud provider API, or upon an external software-defined storage or network
+  control plane.
+
+  For each of these, fill in the following—thinking about running existing user workloads
+  and creating new ones, as well as about cluster-level services (e.g. DNS):
+  - [Dependency name]: installation of csi-external-health-monitor-controller and csi-external-health-monitor-agent sidecars
+    - Usage description:
+      - Impact of its outage on the feature: Installation of csi-external-health-monitor-controller and csi-external-health-monitor-agent sidecars are required for the feature to work. If csi-external-health-monitor-controller is not installed, abnormal volume conditions will not be reported as events on PVCs. Similarly, if csi-external-health-monitor-agent is not installed, abnormal volume conditions will not be reported as events on Pods.
```

Comment:
can you clarify what provides the new API described here: https://github.com/kubernetes/enhancements/blob/master/keps/sig-storage/1432-volume-health-monitor/README.md#csi-changes . It seems as though that component needs to be present and up to date to function.

Reply:
Those are not Kubernetes API changes.  They are in the CSI spec v1.3.   https://github.com/kubernetes-csi/external-health-monitor/blob/master/go.mod#L6

Reply:
> Those are not Kubernetes API changes. They are in the CSI spec v1.3. https://github.com/kubernetes-csi/external-health-monitor/blob/master/go.mod#L6

So whatever provides that API surface needs to support or the csi-external-health-monitor-agent won't be able to query it, right?  I'm imagining a skew case of old node and kubelet with new controller.

Reply:
CSI driver needs to be updated to implement volume health RPCs in controller/node plugins.  The minimum kubernetes version should be 1.13: https://kubernetes-csi.github.io/docs/introduction.html#kubernetes-releases. K8s v1.13 is the minimum supported version for CSI driver for work (of course, different CSI drivers have different requirements on supported k8s versions as well).  Let's say the CSI node plugin on one node has been upgraded to support volume health while it is not upgraded on 3 other nodes, then we will only expect to see volume health events on pods running on that one upgraded node.